### ISSUE TITLE
Multipage Support

### ISSUE THREAD

**Issue by huntcsg** [adamhajari/spyre#10]
I am working on adding support for a 'multi page' app, that would allow for defining routes to different spyre "apps".

You can check out my progress at my fork/master. I've basically got it working. I've got a draft of an example app (that is the example_show_all_the_inputs.py, with 2 additional app classes that are basically exact copies of the first.) At the bottom you can see what I've got for how you would use this functionality.

I added an additional class in server.py that I'm calling Site, that takes the top node as a required argument and exposes a methods to add a new page/route, to change the top node (which would be somewhat dangerous, but there might be a use case to do so), and to launch the site.

I'm not really happy with the naming conventions, as I'm calling the top node the "root" node, which is a little confusing because the object is a server.Root object. Anyone have any input on a better naming convention for a page/route/node and for the top of the page/node/route tree?

There is also an issue with this setup that I think has something to do with how the spinning_wheel is implemented. While the spinning wheel is active the terminal throws a warning/error "QPixmap: It is not safe to use pixmaps outside the GUI thread". My guess is that whatever is being used to determine when the spinning wheel should be replaced by a plot or HTML element is causing this error. It's not strictly causing any problems, but it's probably something that would need to be figured out before incorporating this kind of functionality into your master.

In general what are your thoughts on this?

In terms of how this might be used, I'm not sure if there is anything that could/should be done to the template to allow better use of a multipage site (A navbar across the top? An easy ability to make the control panel buttons be links? an output template/function?). It would be easy enough to just make an HTML output block that would have links, but I'm just wondering if this is something that maybe could be supported better by updating the view.html template.

**Comment by huntcsg** [comment]
Regarding the QPixmap error, it seems that that is not due to the multi-page nature of the app. I'm not sure what is causing it, but it's only happening on my windows machine (when I run the cherrypy server from my windows machine, locally). So, it may just be a platform issue. Or perhaps a python/package version issue.

**Comment by huntcsg** [comment]
I've got an example multipage app running at foolofgod.pythonanywhere.com. In the next couple of days I might build out a more interesting example (than 3 of the same apps). But just for demo purposes, you can find them at:
[foolofgod.pythonanywhere.com/](http://foolofgod.pythonanywhere.com/)
[foolofgod.pythonanywhere.com/other](http://foolofgod.pythonanywhere.com/other)
[foolofgod.pythonanywhere.com/other/another](http://foolofgod.pythonanywhere.com/other/another)

This is all "one" app, running at the root, and then the others are attached to that so the routing works.

**Comment by adamhajari** [comment]
This is definitely something I was hoping to add eventually. Glad to see you've beat me to it. Is it ready to merge?

Right now every spyre app has the control panel on the left side and an output area on the right. I'd eventually like to have several different options for templates. One of those options could be something suitable for an index page linking to several different spyre apps.

As for the QPixmap issue, I do all of my development on Mac and Linux, which is why I missed that. Any interest in looking into it? If not, I'll check out an EC2 instance work on it next week.

**Comment by huntcsg** [comment]
I can put a pull request together in the next couple of days. I am still working on figuring out the best way in terms of how the different apps will be defined by the user.

I will be adding an additional list to templateVars that is a list of apps and their routes. This can then be used in the template to create some sort of navigation.

I started to play around with navigation. I experimented with having a button be made into a link, but I wasn't really happy with that. I started to build a drop down that would sit above the left panel, but I took a detour to do the jquery/javascript tutorials on codeacademy as my experience with javascript is next to null.

So, I'll plan to make a pull request with the nuts and bolts of making multiple apps, and how that would be defined by a user, and then I'll leave it open, how exactly it might be used.
